RFCs/FS-1128-better-interpolated-triple-quoted-strings.md

```diff
@@ -0,0 +1,119 @@
+- [Discussion]()
+
+# Summary
+
+Extend current form of interpolated triple quoted string to allow specifying how many braces should start an interpolation (equal to the number of dollar signs at the start of the literal). For example:
+```
+let json = $$"""
+{
+  "id": "Some guid",
+  "name": "{{user.Name}}",
+};
+"""
+```
+In the above example, `$$` starting the string literal indicates that interpolated expression will be enclosed in `{{...}}` and single `{` and `}` will be treated as content.
+
+# Motivation
+
+Triple quoted strings are already quite handy for embedding e.g. XML in string literals, but not as convenient for embedding some other languages that use `{` and `}` in their syntax, e.g. JSON or CSS. Currently every occurrence of a curly brace in an interpolated string needs to be escaped, which can quickly become tedious. With this extended syntax, only the braces around interpolated expression need special attention, and that is where programmer's focus already is.
+
+The motivation is not to enable any novel use cases, but to make working with interpolated string literals a much more seamless experience in the use cases mentioned above.
+
+This change should:
+1. be backwards compatible with existing support for triple quoted strings
+2. align with C# raw string literals reasonably well
+
+# Detailed design
+
+The existing form of interpolated triple quoted strings: `$"""...{}..."""` is extended to allow multiple dollar signs at the begining of the string literal. The count of these initiatory `$` characters indicates how many `{` and `}` characters are used to delimit interpolation expression within the content of the literal.
+
+Important note: To maintain backward compatibility, there are no changes in semantics for a single-`$` case. This means that it is an edge case in terms of how this feature behaves. It would be nice to have a clean design that covers all cases, but it is not worth breaking backward compatibility for. Therefore the below description applies to the literals starting with 2 or more `$`.
+
+Triple quoted string literals starting with 2 or more `$` characters have no escaping mechanism for `{` or `}`. Instead, the literal can always be constructed in such a way as to ensure that interpolation delimiters will not collide with other curly braces in the content. In a literal that starts with `N` `$` characters, a sequence of consecutive `{` can be at most `2*N-1` characters long and for any sequence of more than `N`, the innermost `N` `{` characters delimit the interpolation, while the outter up to `N-1` are treated as content (analogously for `}`). A sequence of `2*N` or more `{` (or `}`) will result in a compilation error.
+
+Example:
+```
+let text = "3 curly braces delimit interpolation expression"
+$$$"""
+In a string literal with 3x$, { or {{ are simply treated as content
+whereas {{{text}}}.
+It is also allowed to have {, {{, } or }} adjacent to the interpolation delimiters as so:
+{{{{41+1}}}} = {42}
```

Comment:
Suggestion: I’d prefer an example in the style of FSI prompts, so we can more easily follow along and see what the code does?

Reply:
I have split it into single line strings, because it started looking awkward with multiline literal + output. Added FSI style output as `//` comments. Let me know if it looks good now or maybe I should come up with different examples?